src/lib.rs

```diff
@@ -0,0 +1,1607 @@
+        )?;
+
+        let opt_application_stake_balance = ensure_stake_imbalance_matches_staking_policy!(
+            &opt_application_stake_imbalance,
+            &opening.application_staking_policy,
+            AddApplicationError::StakeMissingWhenRequired(StakePurpose::Application),
+            AddApplicationError::StakeProvidedWhenRedundant(StakePurpose::Application),
+            AddApplicationError::StakeAmountTooLow(StakePurpose::Application)
+        )?;
+
+        // Opening is accepting applications
+
+        let (
+            active_stage,
+            applicants,
+            active_application_count,
+            unstaking_application_count,
+            deactivated_application_count,
+        ) = ensure_opening_is_active!(
+            opening.stage,
+            AddApplicationError::OpeningNotInAcceptingApplicationsStage
+        )?;
+
+        ensure_active_opening_is_accepting_applications!(
+            active_stage,
+            AddApplicationError::OpeningNotInAcceptingApplicationsStage
+        )?;
+
+        // Ensure that the new application would actually make it
+        let success = ensure_application_would_get_added!(
+            &opening.application_rationing_policy,
+            &applicants,
+            &opt_role_stake_balance,
+            &opt_application_stake_balance,
+            AddApplicationError::NewApplicationWasCrowdedOut
+        )?;
+
+        //
+        // == MUTATION SAFE ==
+        //
+
+        // If required, deactive another application that was crowded out.
+        if let ApplicationAddedSuccess::CrowdsOutExistingApplication(
+            id_of_croweded_out_application,
+        ) = success
+        {
+            // Get relevant unstaking periods
+            let opt_application_stake_unstaking_period =
+                Self::opt_staking_policy_to_crowded_out_unstaking_period(
+                    &opening.application_staking_policy,
+                );
+            let opt_role_stake_unstaking_period =
+                Self::opt_staking_policy_to_crowded_out_unstaking_period(
+                    &opening.role_staking_policy,
+                );
+
+            // Fetch application
+            let crowded_out_application = <ApplicationById<T>>::get(id_of_croweded_out_application);
+
+            // Initiate actual deactivation
+            //
+            // MUST not have been ignored, is runtime invariant, false means code is broken.
+            // But should we do panic in runtime? Is there safer way?
+            assert_ne!(
+                Self::try_to_initiate_application_deactivation(
+                    &crowded_out_application,
+                    id_of_croweded_out_application,
+                    opt_application_stake_unstaking_period,
+                    opt_role_stake_unstaking_period,
+                    &hiring::ApplicationDeactivationCause::CrowdedOut
+                ),
+                ApplicationDeactivationInitationResult::Ignored
```

Comment:
I don't understand this suggestion. The purpose of this variant is to represent that deactivation was not even initiated because it was in an inapplicable state.